resource/hlapi/bindings/c++/reapi_cli_impl.hpp

```diff
@@ -48,7 +55,70 @@ int reapi_cli_t::match_allocate (void *h, bool orelse_reserve,
                                  const uint64_t jobid, bool &reserved,
                                  std::string &R, int64_t &at, double &ov)
 {
-    return NOT_YET_IMPLEMENTED;
+    resource_context_t *resource_ctx = 
+                    *(static_cast<resource_context_t * *>(h));
```

Comment:
At first glance, this can be unsafe as you are relying on the actual memory layout. Can the caller actually pass the resource context instead of its containing struct pointer?

Reply:
Won't this require reworking the functions so they don't take `void *` as the first argument? That appears to get us back to another version of this discussion above: https://github.com/flux-framework/flux-sched/pull/888#issuecomment-989595644

Reply:
Well, I guess my question is if it is safe to pass `ctx->rqt` (which is `resource_query_t *`) to `void *` and cast it back to `resource_context_t *`.  

```suggestion
    resource_query_t *query = static_cast<resource_query_t *> (h);
```

And then directly use `query` as a handle to get to other state?  

As is, I can't follow the code sorry.

Reply:
Oh, I understand what you mean now.  

Yes, I confess I never understood `*(static_cast<resource_context_t * *>(h))`, but we somehow got that to work in an hour-long hackathon.